fabric-client-gametest-api-v1/src/client/java/net/fabricmc/fabric/impl/client/gametest/TestScreenshotComparisonAlgorithms.java

```diff
@@ -0,0 +1,231 @@
+
+				for (int y = 0; y < needleHeight; y++) {
+					for (int x = 0; x < needleWidth; x++) {
+						int haystackLuminance = haystackData[(needleY + y) * haystackWidth + needleX + x] & 0xff;
+						int needleLuminance = needleData[y * needleWidth + x] & 0xff;
+						sumSquaredDifference += MathHelper.square(haystackLuminance - needleLuminance);
+
+						if (sumSquaredDifference >= threshold) {
+							return false;
+						}
+					}
+				}
+
+				return true;
+			});
+		}
+	}
+
+	public enum Exact implements TestScreenshotComparisonAlgorithm {
+		INSTANCE;
+
+		@Override
+		@Nullable
+		public Vector2i findColor(RawImage<int[]> haystack, RawImage<int[]> needle) {
+			Preconditions.checkNotNull(haystack, "haystack");
+			Preconditions.checkNotNull(needle, "needle");
+
+			int[] haystackData = haystack.data();
+			int[] needleData = needle.data();
+			int haystackWidth = haystack.width();
+			int needleWidth = needle.width();
+			int needleHeight = needle.height();
+
+			return find(haystack, needle, (needleX, needleY) -> {
+				for (int y = 0; y < needleHeight; y++) {
+					for (int x = 0; x < needleWidth; x++) {
+						int haystackColor = haystackData[(needleY + y) * haystackWidth + needleX + x];
+						int needleColor = needleData[y * needleWidth + x];
+
+						if (haystackColor != needleColor) {
+							return false;
+						}
+					}
+				}
+
+				return true;
+			});
+		}
+
+		@Override
+		@Nullable
+		public Vector2i findGrayscale(RawImage<byte[]> haystack, RawImage<byte[]> needle) {
+			Preconditions.checkNotNull(haystack, "haystack");
+			Preconditions.checkNotNull(needle, "needle");
+
+			byte[] haystackData = haystack.data();
+			byte[] needleData = needle.data();
+			int haystackWidth = haystack.width();
+			int needleWidth = needle.width();
+			int needleHeight = needle.height();
+
+			return find(haystack, needle, (needleX, needleY) -> {
+				for (int y = 0; y < needleHeight; y++) {
+					for (int x = 0; x < needleWidth; x++) {
+						byte haystackLuminance = haystackData[(needleY + y) * haystackWidth + needleX + x];
+						byte needleLuminance = needleData[y * needleWidth + x];
+
+						if (haystackLuminance != needleLuminance) {
+							return false;
+						}
+					}
+				}
+
+				return true;
+			});
+		}
+	}
+
+	@Nullable
+	private static Vector2i find(TestScreenshotComparisonAlgorithm.RawImage<?> haystack, TestScreenshotComparisonAlgorithm.RawImage<?> needle, PositionPredicate predicate) {
+		if (needle.width() > haystack.width() || needle.height() > haystack.height()) {
+			return null;
+		}
+
+		return IntStream.rangeClosed(0, haystack.height() - needle.height())
+				.parallel()
+				.mapToObj(needleY -> {
+					int maxNeedleX = haystack.width() - needle.width();
+
+					for (int needleX = 0; needleX <= maxNeedleX; needleX++) {
+						if (predicate.isAt(needleX, needleY)) {
+							return new Vector2i(needleX, needleY);
+						}
+					}
+
+					return null;
+				})
+				.filter(Objects::nonNull)
```

Comment:
There is nothing wrong with this, but if im not mistaken there will be wasted time on all of the none matching y values. The cheap way of doing it would be a an atomic boolean that checks within the for loop if set return null. This saves searching for something that has already been found by another thread.

Edit, just read it again and I think the `findAny()` handles this, close this comment if so.

Reply:
Yeah I believe nonterminal operations in streams are lazy, so it won't generate any more once `findAny()` has found something.

Reply:
Very clever, I dont usually think to use streams to manage parallel tasks like this. 👍